File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

char *check_if_card_num_valid(long num)
{
    int prev_digit = 0;
    int cur_digit = 0;
    int digits = 0;
    int odds_sum = 0;
    int evens_sum = 0;
    int brand_index = 0;

    do
    {
        prev_digit = cur_digit;
        // The current digit
        cur_digit = num % 10;
        // Even digit
        if (digits % 2 == 0)
        {
            evens_sum += cur_digit;
        }
        // Odd digit
        else
        {
            int doubled = cur_digit * 2;
            // Get each bit of the number if the sum is bigger than 10
            // e.g. 12 => 1 + 2
            if (doubled > 9)
            {
                odds_sum += doubled % 10 + doubled / 10;
            }
            else
            {
                odds_sum += doubled;
            }
        }
        digits++;
    }
    while (num /= 10);

    if ((evens_sum + odds_sum) % 10 == 0)
    {
        // Calulate brand index
        brand_index = cur_digit * 10 + prev_digit;
        printf("%d\n", brand_index);
        // VISA
        if (brand_index >= 40
            && brand_index <= 49
            && digits >= 13
            && digits <= 16)
        {
            return "VISA";
        }
        // MasterCard
        else if (brand_index >= 51
                 && brand_index <= 55
                 && digits == 16)
        {
            return "MASTERCARD";
        }
        // American Express
        else if ((brand_index == 34 || brand_index == 37)
                 && digits == 15)
        {
            return "AMEX";
        }
    }

    return "INVALID";
}

int main(void)
{
    long input = get_long("Number: ");

    char *res = check_if_card_num_valid(input);
    printf("%s\n", res);
}

File: pset6/credit/credit.py
from cs50 import get_string

# get credit card number input
input = [int(digit) for digit in get_string('Number: ')]
digits = len(input)

# Odd digit sum
odd_sum = sum(
    (digit * 2 > 9)
    and (digit * 2 % 10 + int(digit * 2 / 10))
    or digit * 2
    for idx, digit in enumerate(input[::-1]) if idx % 2 != 0)

# Even digit sum
even_sum = sum(digit for idx, digit in enumerate(input[::-1]) if idx % 2 == 0)

# Check Brand
if (odd_sum + even_sum) % 10 == 0:
    brand_index = int(''.join(['%d' % i for i in input[0:2]]))
    # Visa
    if 40 <= brand_index <= 49 and 13 <= digits <= 16:
        print('VISA')
    # MasterCard
    elif 51 <= brand_index <= 55 and digits == 16:
        print('MASTERCARD')
    # American Express
    elif brand_index == 34 or brand_index == 37 and digits == 15:
        print('AMEX')
else:
    print('INVALID')